File: shortify_backend/app/schemas/analytics.py
"""Pydantic schemas for analytics responses."""

from pydantic import BaseModel


class ClicksByDate(BaseModel):
    date: str
    clicks: int


class ClicksByCountry(BaseModel):
    country_code: str | None
    clicks: int


class ClicksByBrowser(BaseModel):
    browser: str | None
    clicks: int


class ClicksByOS(BaseModel):
    os: str | None
    clicks: int


class ClicksByDevice(BaseModel):
    device_type: str | None
    clicks: int


class URLAnalyticsResponse(BaseModel):
    slug: str
    total_clicks: int
    unique_visitors: int
    clicks_by_date: list[ClicksByDate]
    clicks_by_country: list[ClicksByCountry]
    clicks_by_browser: list[ClicksByBrowser]
    clicks_by_os: list[ClicksByOS]
    clicks_by_device: list[ClicksByDevice]


class TopURLResponse(BaseModel):
    slug: str
    long_url: str
    title: str | None
    click_count: int
    short_url: str

File: shortify_frontend/src/types/analytics.ts
export interface ClicksByDate {
  date: string
  clicks: number
}

export interface ClicksByCountry {
  country_code: string | null
  clicks: number
}

export interface ClicksByBrowser {
  browser: string | null
  clicks: number
}

export interface ClicksByOS {
  os: string | null
  clicks: number
}

export interface ClicksByDevice {
  device_type: string | null
  clicks: number
}

export interface URLAnalyticsResponse {
  slug: string
  total_clicks: number
  unique_visitors: number
  clicks_by_date: ClicksByDate[]
  clicks_by_country: ClicksByCountry[]
  clicks_by_browser: ClicksByBrowser[]
  clicks_by_os: ClicksByOS[]
  clicks_by_device: ClicksByDevice[]
}

export interface TopURLResponse {
  slug: string
  long_url: string
  title: string | null
  click_count: number
  short_url: string
}
